tstime/rate/value.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package rate
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"tailscale.com/tstime/mono"
+)
+
+// Value measures the rate at which an event occurs,
+// exponentially weighted towards recent activity.
+// It is guaranteed to occupy O(1) memory and operate in O(1) runtime.
+type Value struct {
```

Comment:
add a note that it is not threadsafe

Reply:
same, but a mutex might be too expensive for my application.

Reply:
I went just a trivial `sync.Mutex` for now. The single-threaded performance degrades from 62ns to 70ns. Around 25ns of that is just calling `mono.Now()`.

We can optimize the concurrency aspect of this in a future PR. Any ideas that I can come up greatly complicates the code, so we should review the math portion of this by itself. That itself is a large tricky enough.